fuel-txpool/src/txpool/tests.rs

```diff
@@ -0,0 +1,1023 @@
+        .await
+        .expect("Tx0 should be Ok, got Err");
+    txpool
+        .insert_inner(tx2.clone(), &db)
+        .await
+        .expect("Tx1 should be Ok, got Err");
+    let tx3_result = txpool
+        .insert_inner(tx3.clone(), &db)
+        .await
+        .expect("Tx2 should be Ok, got Err");
+
+    let mut seen = HashMap::new();
+    txpool
+        .dependency()
+        .find_dependent(tx3_result.inserted, &mut seen, txpool.txs());
+
+    let mut list: Vec<ArcPoolTx> = seen.into_iter().map(|(_, tx)| tx).collect();
+    // sort from high to low price
+    list.sort_by_key(|tx| Reverse(tx.price()));
+    assert_eq!(list.len(), 3, "We should have three items");
+    assert_eq!(list[0].id(), tx1.id(), "Tx1 should be first.");
+    assert_eq!(list[1].id(), tx2.id(), "Tx2 should be second.");
+    assert_eq!(list[2].id(), tx3.id(), "Tx3 should be third.");
+}
+
+#[tokio::test]
+async fn tx_at_least_min_gas_price_is_insertable() {
+    let mut rng = StdRng::seed_from_u64(0);
+    let mut txpool = TxPool::new(Config {
+        min_gas_price: 10,
+        ..Default::default()
+    });
+    let db = MockDb::default();
+
+    let (_, gas_coin) = setup_coin(&mut rng, Some(&db));
+    let tx = Arc::new(
+        TransactionBuilder::script(vec![], vec![])
+            .gas_price(10)
+            .add_input(gas_coin)
+            .finalize_as_transaction(),
+    );
+
+    txpool
+        .insert_inner(tx, &db)
+        .await
+        .expect("Tx should be Ok, got Err");
+}
+
+#[tokio::test]
+async fn tx_below_min_gas_price_is_not_insertable() {
+    let mut rng = StdRng::seed_from_u64(0);
+    let mut txpool = TxPool::new(Config {
+        min_gas_price: 11,
+        ..Default::default()
+    });
+    let db = MockDb::default();
+
+    let (_, gas_coin) = setup_coin(&mut rng, Some(&db));
+    let tx = Arc::new(
+        TransactionBuilder::script(vec![], vec![])
+            .gas_price(10)
+            .add_input(gas_coin)
+            .finalize_as_transaction(),
+    );
+
+    let err = txpool
+        .insert_inner(tx, &db)
+        .await
+        .expect_err("expected insertion failure");
+    assert!(matches!(
+        err.root_cause().downcast_ref::<Error>().unwrap(),
+        Error::NotInsertedGasPriceTooLow
+    ));
+}
+
+#[tokio::test]
+async fn tx_inserted_into_pool_when_input_message_id_exists_in_db() {
+    let (message, input) = create_message_predicate_from_message(5000, None);
+
+    let tx = Arc::new(
+        TransactionBuilder::script(vec![], vec![])
+            .add_input(input)
+            .finalize_as_transaction(),
+    );
+
+    let mut db = MockDb::default();
+    db.storage::<Messages>()
+        .insert(&message.id(), &message)
+        .unwrap();
+    let mut txpool = TxPool::new(Default::default());
+
+    txpool
+        .insert_inner(tx.clone(), &db)
+        .await
+        .expect("should succeed");
+
+    let tx_info = TxPool::find_one(&RwLock::new(txpool), &tx.id())
```

Comment:
This `RwLock` in the `TxPool` looks soooooo strange=D I hope we will refactor it soon=)

Reply:
Also worth looking into!